microbench/src/main/java/io/netty/handler/codec/mqtt/MqttMessageTypeValueOfBench.java

```diff
@@ -0,0 +1,175 @@
+            5, 2, 14, 3, 11, 5, 4, 14, 14, 2, 7, 7, 2, 3, 11, 2, 10, 9, 13, 3, 4, 2, 10, 1, 2, 10, 7, 7, 6, 8, 8, 12,
+            14, 8, 13, 1, 9, 5, 9, 1, 14, 2, 5, 5, 5, 3, 13, 11,
+            9, 6, 11, 1, 10, 13, 4, 7, 9, 6, 3, 4, 11, 8, 13, 3, 13, 12, 7, 7, 5, 9, 11, 3, 9, 6, 5, 6, 6, 11, 9, 2, 7,
+            1, 12, 7, 5, 8, 11, 4, 9, 10, 11, 12, 7, 8, 1, 2,
+            3, 14, 3, 9, 11, 9, 7, 4, 4, 4, 8, 4, 4, 2, 5, 8, 2, 11, 7, 13, 2, 14, 3, 6, 7, 14, 12, 6, 9, 11, 10, 9, 6,
+            10, 6, 14, 4, 1, 7, 12, 4, 13, 10, 2, 2, 3, 3, 14,
+            14, 2, 9, 12, 3, 9, 7, 6, 12, 8, 9, 5, 11, 13, 14, 14, 4, 1, 11, 14, 5, 9, 7, 14, 7, 13, 7, 14, 3, 14, 2, 8,
+            2, 5, 10, 12, 14, 9, 11, 3, 14, 8, 12, 12, 5, 2,
+            6, 2, 1, 14, 12, 8, 14, 1, 11, 14, 8, 9, 9, 1, 12, 13, 7, 8, 10, 5, 8, 5, 14, 13, 14, 3, 14, 2, 9, 12, 3,
+            10, 3, 2, 4, 3, 5, 5, 10, 10, 13, 10, 7, 6, 4, 2, 10,
+            8, 14, 2, 7, 1, 2, 7, 13, 2, 3, 6, 14, 3, 8, 12, 3, 4, 12, 6, 3, 10, 6, 14, 9, 1, 6, 3, 14, 7, 1, 7, 2, 12,
+            9, 5, 9, 6, 13, 5, 11, 13, 11, 10, 1, 14, 9, 13, 8,
+            12, 14, 14, 8, 13, 2, 6, 14, 2, 2, 9, 12, 9, 7, 2, 11, 4, 6, 8, 10, 12, 10, 11, 2, 9, 9, 5, 4, 3, 4, 4, 10,
+            3, 1, 12, 13, 9, 8, 1, 9, 9, 4, 2, 7, 3, 4, 11, 11,
+            8, 10, 14, 5, 14, 1, 10, 10, 13, 5, 6, 13, 14, 5, 7, 11, 4, 13, 3, 14, 7, 2, 10, 13, 2, 4, 14, 5, 1, 12, 3,
+            13, 11, 2, 11, 14, 2, 5, 8, 13, 4, 13, 13, 3, 3,
+            3, 13, 6, 11, 5, 3, 2, 13, 9, 2, 10, 8, 3, 11, 4, 6, 12, 14, 6, 2, 14, 1, 2, 6, 8, 4, 12, 8, 11, 9, 1, 7, 1,
+            10, 4, 10, 9, 9, 3, 11, 5, 10, 8, 9, 4, 13, 4,
+            5, 7, 12, 14, 12, 6, 1, 2, 10, 9, 10, 12, 1, 2, 6, 9, 5, 13, 4, 6, 11, 7, 1, 3, 10, 2, 1, 13, 14, 3, 5, 5,
+            5, 7, 14, 9, 9, 3, 12, 1, 1, 1, 3, 12, 6, 9, 7, 8, 1,
+            8, 2, 8, 13, 1, 11, 11, 1, 4, 10, 4, 3, 10, 3, 2, 2, 8, 2, 4, 13, 14, 4, 12, 14, 7, 6, 7, 13, 7, 11, 13, 12,
+            14, 1, 14, 3, 4, 13, 12, 10, 5, 12, 12, 4, 5, 6,
+            9, 12, 13, 3, 4, 13, 8, 14, 3, 2, 8, 5, 6, 13, 8, 7, 4, 5, 8, 14, 8, 14, 7, 5, 4, 9, 12, 12, 10, 3, 1, 12,
+            5, 1, 11, 6, 10, 5, 14, 4, 5, 13, 8, 11, 13, 4, 9,
+            9, 7, 6, 2, 2, 5, 12, 13, 13, 6, 11, 13, 12, 10, 6, 7, 1, 2, 6, 1, 9, 10, 14, 7, 9, 2, 2, 2, 8, 8, 11, 14,
+            12, 9, 13, 1
+    };
+
+    int[] types;
+    int mask;
+    long next;
+
+    @Setup
+    public void initDataSet() {
+        this.types = DATASET;
+        if (!Pow2.isPowerOfTwo(DATASET.length)) {
+            System.err.println("the DATASET length should be power of 2 sized!");
+            System.exit(-1);
+        }
+        this.mask = DATASET.length - 1;
+        this.next = 0;
+    }
+
+    @Benchmark
+    public MqttMessageType getViaArray() {
+        long next = this.next;
+        int nextIndex = (int) (next & mask);
+        MqttMessageType type = MqttMessageTypeArray.valueOf(types[nextIndex]);
```

Comment:
This is going to cause a lot of contention on this.next. Why not just use a for loop and iterate over all of them, instead of trying to do measure just a single one?

Reply:
I've tried that here: https://github.com/netty/netty/pull/10400#issuecomment-658059392
And the results gotten were quite similar: https://github.com/netty/netty/pull/10400/commits/b30c74b06f5720911efc7b53f846ad470091ff30 instead you can check the "wrong" version of the the benchmark on https://github.com/netty/netty/pull/10400#issuecomment-658059392, because not using BlackHole on each iteration can cause the for loop to perform some loop unrolling, making the `getViaArray` to be better then it is.

I'm still not 100% sure if I should use https://github.com/netty/netty/pull/10400#issuecomment-658059392 or this benchmark (using the & trick) TBH...

> his is going to cause a lot of contention on this.next.

wdym with contention?

Reply:
I think your approach with the blackhole.consume() in the for loop is probably the right one, in that it shows that when you iterate over multiple values (and you consume the results) then the lookup through the array is faster than looking up with the switch.

This approach (reading/writing the local state) might end up being an unnecessary overhead, and you'll have a lot of write traffic happening on each loop. I still don't see the benefit in trying to benchmark each lookup individually, when you can lookup in a for loop (with bh.consume) and it will give you the data that suggests array lookups are faster than switch statements.

Reply:
Yep, agree: I've convinced me: trying to get numbers for the single case isn't meaningful for the purpose of the benchmark and it just create unneeded noise in the results :+1 
Let me send a new commit with very last iteration of the bench, thanks for good advices :)

Reply:
@alblue if the state scope it `Thread` there should be no contention. The risk with writing loops into benchmarks is that either:
1. You'll get loop unrolling
2. You'll get OSR compilations

Resulting in confusing results. I have used the "one element" benchmark in the past with no issues. The overhead of elements load is very small and easily measurable, so I would generally go for that method rather than the loop.

A further point for consideration is using the "alternative" blackhole method (a `dontinline` empty method), which might be appropriate given the very small operation we are examining here.

I think there's no point to invest further in measurement here, but commenting for feedback/clarity.

Reply:
@nitsanw 
> You'll get OSR compilations

urgh, now I've remembered...this one was the main reason why I've switched to the single iteration method, you're right